clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp

```diff
@@ -4220,6 +4220,51 @@ class CIRAbsOpLowering : public mlir::OpConversionPattern<mlir::cir::AbsOp> {
     return mlir::success();
   }
 };
+class CIRSignBitOpLowering
+    : public mlir::OpConversionPattern<mlir::cir::SignBitOp> {
+public:
+  using OpConversionPattern<mlir::cir::SignBitOp>::OpConversionPattern;
+
+  mlir::LogicalResult
+  matchAndRewrite(mlir::cir::SignBitOp op, OpAdaptor adaptor,
+                  mlir::ConversionPatternRewriter &rewriter) const override {
+    assert(!::cir::MissingFeatures::isPPC_FP128Ty());
+    int width = getPrimitiveSizeInBits(op.getInput().getType());
+    auto intTy = mlir::IntegerType::get(rewriter.getContext(), width);
+    auto bitcast = rewriter.create<mlir::LLVM::BitcastOp>(op->getLoc(), intTy,
+                                                          adaptor.getInput());
+    auto zero = rewriter.create<mlir::LLVM::ConstantOp>(op->getLoc(), intTy, 0);
+    auto cmpResult = rewriter.create<mlir::LLVM::ICmpOp>(
+        op.getLoc(), mlir::LLVM::ICmpPredicate::slt, bitcast.getResult(), zero);
+    auto converted = rewriter.create<mlir::LLVM::ZExtOp>(
+        op.getLoc(), mlir::IntegerType::get(rewriter.getContext(), 32),
+        cmpResult);
+    rewriter.replaceOp(op, converted);
+    return mlir::success();
+  }
+
+private:
+  static int getPrimitiveSizeInBits(mlir::Type ty) {
```

Comment:
Instead of creating this function, you should query the datalayout directly. See uses of `mlir::DataLayout dataLayout(module);` in LowerToLLVM.cpp for more examples.

Reply:
This patch is blocked by https://github.com/llvm/clangir/issues/1057. Please review https://github.com/llvm/clangir/pull/1058 first.